KEEP/initramfs.sh

```diff
@@ -57,12 +57,35 @@ mopts=""
 dev=$(findfs "$root")
 mount $mopts "$dev" /root || rescue "rootfs: mount failed with code $?"
 
+# Mount the overlayfs
+if [ -n "$overlay" ] || [ -n "$overlayfstype" ]; then
+	# Move the existing root (new rom) away
+	mkdir -p /rom /overlay
+	mount --move /root /rom
+
+	# Mount the upper (writable) source
+	mopts=""
+	[ -n "$overlayflags" ]  && mopts="$mopts -o $overlayflags"
+	[ -n "$overlayfstype" ] && mopts="$mopts -t $overlayfstype"
+	mount $mopts `findfs "$overlay"` /overlay || rescue "overlayfs: mount failed with code $?"
+
+	# Data is for overwritten data, work is a tempdir for atomic renames by overlayfs
+	mkdir -p /overlay/data /overlay/work
+
+	# Mount the rom and the writable overlay together on /root
+	mount -t overlay overlay -olowerdir=/rom,upperdir=/overlay/data,workdir=/overlay/work /root \
+		|| rescue "overlaymount failed with code $?"
+fi
+
 # Check for some misc. conditions
-[ -x "/root$init" ] || rescue "$init not found"
 [ -n "$rescue" ] && rescue "Rescue shell requested by kernel options"
+[ -x "/root$init" ] || rescue "rootfs: $init not found"
+
+# Inherit to rootfs
+for dir in dev rom overlay boot; do [ -d /$dir ] && mkdir -p /root/$dir && mount --move /$dir /root/$dir; done
+
+# Umount and forget
+for dir in sys proc; do [ -d /$dir ] && umount /$dir; done
```

Comment:
If /boot gets mounted, and an file in /boot gets loopmounted (e.g. as new root), it prevents the /boot from being able to be umounted. Even worse, the loopmounted image will be existant in a directory tree outside of the actual systems directory root (i call this shadow mount). For the same reason /rom and /overlay are getting moved to the new rootfs